### PR TITLE
ENH: Peaks declustering and signal up/down-crossing finding

### PR DESCRIPTION
**Pull request by ali-cetin-4ss**
This pull request adds peaks declustering and basic signal crossing detection similar to argrelmax, argrelmin.
The pull request is feature complete with unit-tests and doc.

Note: This PR is a duplicate of https://github.com/scipy/scipy/pull/9397 . PR #9397 was abandoned due to git mess. (cc: @tylerjereddy , @ilayn , @lagru )

### CODE REVIEW

**Comment by ace-e4s** [comment]
The CI fails with this:

`
  
============
scipy.signal
============

Non-deprecated objects in __all__: 180
Objects in refguide: 156

ERROR: objects in scipy.signal.__all__ but not in refguide::

    argcross
    argcrossdown
    argcrossup
    decluster_peaks

This issue can be fixed by adding these objects to
the function listing in __init__.py for this module

ERROR: refguide or doctests have errors`

What is this?

**Comment by larsoner** [comment]
> This issue can be fixed by adding these objects to
> the function listing in `__init.py__` for this module

This is the hint. You need to add new public functions to the list at the top of `scipy/signal/__init__.py`

**Comment by ace-e4s** [comment]
Ahh, I see, thanks @larsoner.  But since this PR introduces a new submodule, maybe one of the more senior members should suggest a name/layout.

**Comment by ace-e4s** [comment]
I have created a notebook that demonstrates the features included in this pull-request. It ought to make a review easier. https://github.com/ace-e4s/notebooks/blob/master/Peak%20declustering%20and%20signal%20crossings%20detection.ipynb

I hope this is helpful. Cheers.

**Comment by rgommers** [comment]
Thanks @ace-e4s, adding a notebook is indeed helpful. I'll tag this for the next release, so we get a reminder to get back to it and at least make a decision. For reviewers, this is the mailing list thread on adding this feature:  https://mail.python.org/pipermail/scipy-dev/2018-October/023127.html

File: scipy/signal/_crossing_finding.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    comparator_1, comparator_2 = _select_cross_comparator(cross_type)
+
+    arr = data.take(locs[:-1], axis=axis, mode=mode)
+    cond = comparator_1(arr, threshold)
+    data.take(locs[1:], axis=axis, mode=mode, out=arr)
+    cond &= comparator_2(arr, threshold)
+    return np.concatenate([cond, padding], axis=axis)
+
+
+def argcrossup(x, threshold=None, axis=0, mode='clip'):
```
Comment (tylerjereddy) [inline scipy/signal/_crossing_finding.py]: I think this was brought up in the mailing list thread, but there is some obvious concern that something like `scipy.signal.argrelextrema` could be used to get something quite similar.

Also, what about things like `np.diff(np.sign(data - threshold))`? That seems like it has intuitive potential to do something similar (maybe with slight modification) & potentially with excellent performance.

File: scipy/signal/_peak_finding.py [hunk 2]
```diff
@@ -1297,3 +1298,158 @@ def find_peaks_cwt(vector, widths, wavelet=None, max_distances=None,
...
+
+def _merge_clusters(peaks_above, peaks_below, runs=0):
+    """
+    Generator that merges naive peak clusters according to some declustering
+    rule.
+
+    Parameters
+    ----------
+    peaks_above : sequence
+        Sequence of arrays with indecies of peaks above a thershold within a
```
Comment (tylerjereddy) [inline scipy/signal/_peak_finding.py]: typo "indecies" & below as well

File: scipy/signal/_peak_finding.py [hunk 3]
```diff
@@ -1297,3 +1298,158 @@ def find_peaks_cwt(vector, widths, wavelet=None, max_distances=None,
...
+
+    Runs (``runs``) declustering is as following:
+     1. Identify clusters of exceedences, i.e., find all peaks between two
+        upcrossings above the threshold (cf. ``x_th``).
+     2. Merge clusters that are seperated with less than k runs (cf. ``runs``).
+        The number of peaks below the threshold seperating a down-crossing and
+        the subsequent up-crossing is called runs.
+     3. Select only the n-th largest peaks from each cluster (cf. ``order``).
+
+    The difference between the methods are subtle, and they even overplap for
```
Comment (tylerjereddy) [inline scipy/signal/_peak_finding.py]: typo "overplap"

File: scipy/signal/_peak_finding.py [hunk 4]
```diff
@@ -1297,3 +1298,158 @@ def find_peaks_cwt(vector, widths, wavelet=None, max_distances=None,
...
+    --------
+    find_peaks_cwt
+        Find peaks using the wavelet transformation.
+    find_peaks
+        Find peaks inside a signal based on peak properties.
+
+    """
+    x = _arg_x_as_expected(x)
+
+    if x_th is None:
```
Comment (tylerjereddy) [inline scipy/signal/_peak_finding.py]: and `np.nan`  handling?

File: scipy/signal/tests/test_peak_finding.py [hunk 5]
```diff
@@ -21,11 +21,25 @@
...
+    _identify_ridge_lines,
+    _merge_clusters,
+    decluster_peaks
 )
 from scipy.signal._peak_finding_utils import _local_maxima_1d, PeakPropertyWarning
 
 
+def _gen_gaussian_signal(t, freq_range, freq_div):
+    """Generate (white) Gaussian signal."""
+    amp = np.random.randn(freq_div)
```
Comment (tylerjereddy) [inline scipy/signal/tests/test_peak_finding.py]: may need to pin a seed here if it isn't happening somewhere else in module?

**Review by tylerjereddy — COMMENTED** [review]
I started reviewing this & added a number of comments / thoughts. Some are more meaningful / API & algorithm related, and others less so.

Anyway, I'm not sure I have enough confidence to make the final decision here--the relationship between the new public functionality exposed here & things like `argrelmax` (which you mention) & the NumPy tricks I mention in the review are perhaps points of modest hesitation / discussion before merging.

**Comment by tylerjereddy** [comment]
I'll bump the milestone here, but hopefully my thoughts / revisions encourage a decision on the appropriate direction of this PR.

**Comment by ace-e4s** [comment]
Bumping up the milestone is reasonable considering how close the planned 1.3 release is. I'll go through the comments and fix the apparent ones. After the 1.3 release, we can pick up the discussion considering the direction this PR. Anyway, thanks for the comments and good luck with the 1.3 release @tylerjereddy

**Comment by larsoner** [comment]
@ace-e4s `find_peaks` has a bunch of ways of controlling which peaks get returned. Could declustering be included there?

After reading through the mailing list a bit, I'd give a +0.75 that this is in scope for SciPy. I am no peak finding / analysis expert but it seems relevant and useful. Given that IIUC @rgommers was on the fence, I'm mostly for it, and nobody thus far objects to its inclusion, I say we go for it. @rgommers it would be good to hear if you agree before @ace-e4s you work on it further, though.

@ace-e4s if you do get the green light, the first thing is probably to rebase so the CIs run properly again. Then I can review what you have so far.

**Comment by lagru** [comment]
> @ace-e4s find_peaks has a bunch of ways of controlling which peaks get returned. Could declustering be included there?

Yes. `decluster_peaks` uses `_local_maxima_1d` under the hood and basically selects a subset of peaks based on conditions (same as `find_peaks`). However given the number of parameters and length of explanation I'd argue that the documentation would be more contained and easier to grasp if it is inside its own docstring. And there might be confusion between the different "thresholds" used in both functions. So it's a trade off.

@ace-e4s would the algorithm still work/make sense if it were run on a subset of peaks? In that case it might be useful to add an input parameter `peaks` to give the user more control and facilitate interoperability with other peak finding functions. `_local_maxima_1d` would then function as a backup in case the user doesn't provide peaks himself.

**Comment by rgommers** [comment]
> After reading through the mailing list a bit, I'd give a +0.75 that this is in scope for SciPy. I am no peak finding / analysis expert but it seems relevant and useful. Given that IIUC @rgommers was on the fence, I'm mostly for it, and nobody thus far objects to its inclusion, I say we go for it. @rgommers it would be good to hear if you agree before @ace-e4s you work on it further, though.

I'm in principle in favor of better peak finding in `scipy.signal`. I'm still uncomfortable with both what we already have, and what's proposed here though. Without a reference paper and without it being an improvement over what's available already in Python (see e.g.https://github.com/MonsieurV/py-findpeaks), it's unclear to me if this is helpful or not. And I do mean it's unclear, not that I'm negative - I just can't tell either way.

Example from the notebook linked to above; why does this make sense?

image

**Comment by ace-e4s** [comment]
`find_peaks` and `decluster_peaks` differ "conceptually". `find_peaks` selects a subset of peaks based on deterministic criteria, whereas `decluster_peaks` does the same based on stochastic concerns. The purpose of "peak declustering" is to take a sequence of statistically dependent peaks, and then try to select a subset that is (hopefully) statistically independent so that simple statistical analysis can be used. The methods are admittedly empirical, but accepted approaches in the field of extreme value analysis. As reference: "An introduction to statistical modelling of extreme values", S. Coles, 2004, Springer, pp. 92-104 (Chap: Extremes of dependent sequences - in particular see example 5.3.3 and 5.3.4).  

Also, see here for an explanation of a possible implementation: https://www.orcina.com/SoftwareProducts/OrcaFlex/Documentation/Help/Content/html/Extremevaluestatisticstheory.htm

**Comment by ace-e4s** [comment]
> Example from the notebook linked to above; why does this make sense?

@rgommers , I agree that it looks weird, but there is a logic here. Lets review the criteria in the notebook for this example:

- order=3 -> peak the 3 largest peaks in a cluster
- runs=2 -> there must be at least 2 peaks below the threshold to separate two clusters

Now lets review the peaks. The 1st peak belongs to separate cluster, but its companion peaks are not visible in the plot. Note that there are 2 peaks below the threshold between 1st and 2nd peak. Therefore, 2nd, 3rd, and 4th peaks are 3 largest peaks in their cluster. Note that we dont observe 2 consecutive peaks below a peak after the first time.

So to summarize, we see 2 clusters in the plot. The first cluster is observed only partially. For the second cluster, we see all of the peaks we select, but have not observed an termination of the cluster yet.

Also, most common use case will be setting order=1.

**Comment by rgommers** [comment]
Thanks for the explanations @ace-e4s. 

I think the examples when zooming in (as in the image above) are perhaps a little misleading. It looks like a periodic signal with noise added. Where a good fraction or all of the peaks are detected. Extreme value analysis on the other hand seems focused on _extreme_ outliers (e.g. floods, wildfires, etc. - see https://en.wikipedia.org/wiki/Extreme_value_theory). As such it's more about statistics of very unusual events than it is about peak detection in signal processing.

This PR contains two separate sets of functionality. The up/down-crossings seems to fit in `scipy.signal` and looks sensible to me. The "extreme value" part I'm not so sure about.

**Comment by rgommers** [comment]
makes sense, let's try to get the up/down crossing into 1.4.0

**Comment by e-q** [comment]
As Tyler mentioned above, variations on `np.diff(np.sign(data - threshold))` can get things done.

I think the scenario where a distinct function makes sense for `scipy.signal` is if it is part of a broader set of functions for dealing with digital/pulsed waveforms. These could be things like duty cycle, pulse width distribution, pulse rate, etc. Has there been any demand for these kinds of functions? I think it could make sense, if someone wants to put it together as a cohesive unit.

Given that 1.4.0 is branching soon, I'm in favor of letting this slide for the moment.

**Comment by larsoner** [comment]
We are getting too close to branching at this point, so I removed the 1.4.0 label.

**Comment by ace-e4s** [comment]
Although I would loved to update the pr, I'm swamped with other thing at the moment. But I'll try to make the agreed changes before the end of this year.

**Comment by ace-e4s** [comment]
I see that a new release is close. To be realistic, I don't have the time to fix this PR. And the scope ended up to be pretty limited anyways. So, I suggest we can close it... Thanks for all the feedback guys, keep up the great work!

**Comment by larsoner** [comment]
Thanks @ace-e4s , we can reopen at some point if need be
